Checkout: Use ResponseCartProduct for WPLineItem

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

The `WPCOMCartItem` type was created as a sort of hack to allow passing `ResponseCartItem` object data through checkout via the `useLineItems` hook that's part of the `composite-checkout` package, since the generic product data available in that hook's types is not sufficient. However, now that `useShoppingCart` is available, we can get the `ResponseCartItem` products directly where they are needed. This will make it easier to manipulate and query cart items in the code since there will only be one data type we need to understand, and since `ResponseCartItem` predates `WPCOMCartItem`, there's already considerable support for it in calypso.

This PR refactors the `WPLineItem` (aka: `LineItem`) component in calypso checkout so that it uses a `ResponseCartProduct` instead of a `WPCOMCartItem`. Because this component was also being used for non-product line items like total, subtotal, tax, coupons, and credits, this PR also creates a `NonProductLineItem` component for those.

There's still more work to be done to remove `WPCOMCartItem` entirely, but this will take care of the biggest use of that type: the checkout products list.

In order to be certain that these types were being used correctly, this PR also converts the whole `WPOrderReviewLineItems` file to TypeScript, which resulted in this PR being quite large. It's probably easiest to review the commits one-by-one than all together.

#### Screenshots
Product list:
<img width="493" alt="Screen Shot 2021-02-08 at 5 43 55 PM" src="https://user-images.githubusercontent.com/2036909/107291050-8c06d100-6a35-11eb-9ece-1cfd161e172e.png">

Product variation picker:
<img width="495" alt="Screen Shot 2021-02-08 at 5 44 02 PM" src="https://user-images.githubusercontent.com/2036909/107291058-8f01c180-6a35-11eb-8528-c09021f909a1.png">

Tax, credits, subtotal on last step:
<img width="500" alt="Screen Shot 2021-02-08 at 5 44 35 PM" src="https://user-images.githubusercontent.com/2036909/107291071-945f0c00-6a35-11eb-98ba-6dda8597de48.png">


#### Testing instructions

- Make sure your account has a little bit of money as free credits (not enough to pay for a whole plan).
- Add a plan to your cart and visit checkout.
- Verify that the plan title and price are displayed.
- Verify that the plan's original price is displayed crossed-out. (If you don't have a discount, try to get one somehow to make sure this works.)
- Click "Edit" on the first step and verify that the product variation picker is displayed.
- For the contact form, use a postal code and country that require taxes (eg: 10001, US).
- Verify that the subtotal, total, credits, and taxes are displayed in the last step.
- Add a coupon; verify that it was added successfully and displays with the order.
- Remove the coupon; verify that it was removed successfully.

sirbrillig committed:
Extract creation of product label into a function

sirbrillig committed:
Begin replacing WPCOMCartItem with ResponseCartProduct in WPLineItem

sirbrillig committed:
Make LineItemPrice not depend on WPCOMCartItem

This alters LineItemPrice so that it takes only the data it needs to
render the price, which must be calculated ahead of time, removing its
dependency on the WPCOMCartItem data type.

sirbrillig committed:
Modify GSuiteUsersList/TitanMailMeta to use ResponseCartProduct

This updates the prop type for each of these components to replace
WPCOMCartItem with ResponseCartProduct.

sirbrillig committed:
Convert LineItemSublabelAndPrice to use ResponseCartProduct

sirbrillig committed:
Update DiscountCallout components to use ResponseCartProduct

sirbrillig committed:
Replace WPCOMCartItem with ResponseCartProduct in ItemVariationPicker

sirbrillig committed:
Add premium and new_quantity to ResponseCartProductExtra

sirbrillig committed:
Make payload optional in onEvent

sirbrillig committed:
Convert WPOrderReviewLineItems to TypeScript

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx:
```diff
@@ -456,133 +346,172 @@ function TitanMailMeta( { item, isRenewal } ) {
+				title: String( translate( 'You are about to remove your plan from the cart' ) ),
+				description: String(
+					createUserAndSiteBeforeTransaction
+						? translate( 'When you press Continue, we will remove your plan from the cart.' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'When you press Continue, we will remove your plan from the cart and your site will continue to run with its current plan.' )` *ES Score: 6*

sirbrillig replied:
I don't know why these strings were not translated in the first place when they were added, but this is different because the site does not exist yet.

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx:
```diff
@@ -456,133 +346,172 @@ function TitanMailMeta( { item, isRenewal } ) {
+				title: String( translate( 'You are about to remove your product from the cart' ) ),
+				description: String(
+					createUserAndSiteBeforeTransaction
+						? translate( 'When you press Continue, we will remove your product from the cart.' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'When you press Continue, we will remove your product from the cart and your site will continue to run without it.' )` *ES Score: 6*

sirbrillig replied:
I don't know why these strings were not translated in the first place when they were added, but this is different because the site does not exist yet.

sirbrillig committed:
Convert styled components to TS in WPOrderReviewLineItems

sirbrillig committed:
Add getTotalLineItem to create LineItem for total

sirbrillig committed:
Make separate component for non-product line items

This way WPLineItem does not need to handle rendering the line item rows
for "total", "credits", "subtotal", or "tax".

sirbrillig committed:
Use NonProductLineItem for rendering non-products

sirbrillig committed:
Add getTaxLineItem

sirbrillig committed:
Use getTaxLineItem to display taxes in PaymentMethodStep

sirbrillig committed:
Add getCreditsLineItem

sirbrillig committed:
Add getSubtotalLineItem

sirbrillig committed:
get subtotal and credits line items directly in PaymentMethodStep

sirbrillig committed:
Remove unused props from styled WPLineItem

sirbrillig committed:
Add isSummary to WPNonProductLineItem

sirbrillig committed:
Allow WPOrderReviewLineItems to render LineItemType also

sirbrillig committed:
Add getCouponLineItem

sirbrillig committed:
Use useShoppingCart for products in WPOrderReviewLineItems

nbloomf reviewed (approved):
Looks good and works as described. The TS conversion here is a nice bonus. :)

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5507586

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5507586).
